src/control/server/server.go

```diff
@@ -207,270 +101,235 @@ func Start(log *logging.LeveledLogger, cfg *config.Server) error {
 		control.WithConfig(cliCfg),
 		control.WithClientLogger(log))
 
-	// Create event distributor.
+	// Create event distribution primitives.
 	eventPubSub := events.NewPubSub(ctx, log)
-	defer eventPubSub.Close()
 
-	// Init management RPC subsystem.
-	mgmtSvc := newMgmtSvc(harness, membership, sysdb, rpcClient, eventPubSub)
+	// Create storage subsystem providers.
+	scmProvider := scm.DefaultProvider(log)
+	bdevProvider := bdev.DefaultProvider(log)
 
-	// Forward published actionable events (type RASTypeStateChange) to the
-	// management service leader, behavior is updated on leadership change.
-	eventForwarder := control.NewEventForwarder(rpcClient, cfg.AccessPoints)
-	eventPubSub.Subscribe(events.RASTypeStateChange, eventForwarder)
-	// Log events on the host that they were raised (and first published) on.
-	eventLogger := control.NewEventLogger(log)
-	eventPubSub.Subscribe(events.RASTypeAny, eventLogger)
+	return &server{
+		log:          log,
+		cfg:          cfg,
+		faultDomain:  fd,
+		harness:      harness,
+		membership:   membership,
+		sysdb:        sysdb,
+		pubSub:       eventPubSub,
+		evtForwarder: control.NewEventForwarder(rpcClient, cfg.AccessPoints),
+		evtLogger:    control.NewEventLogger(log),
+		ctlSvc:       NewControlService(log, harness, bdevProvider, scmProvider, cfg, eventPubSub),
+		mgmtSvc:      newMgmtSvc(harness, membership, sysdb, rpcClient, eventPubSub),
+		scmProvider:  scmProvider,
+		bdevProvider: bdevProvider,
+	}, nil
+}
 
-	var netDevClass uint32
+func (srv *server) shutdown() {
+	srv.pubSub.Close()
+}
 
-	netCtx, err := netdetect.Init(context.Background())
+// initNetwork resolves local address and starts TCP listener, initializes net
+// detect library and warns if configured number of engines is less than NUMA
+// node count.
+func (srv *server) initNetwork(ctx context.Context) (context.Context, func(), error) {
```

Comment:
I think it would be better to confine all interaction with the netdetect package to this method, if possible. So you can initialize the library, get your special context, then iterate through the engine configs and do all validation/adjustments in one place. This has a few advantages: 1) No need to "leak" the netdetect context back out into the rest of the server, 2) Provides a single focus area for all of the network-y stuff rather than sprinkling it around, 3) Allows us to fail faster in the event that something is wrong with the config/environment.